base/bucket_gocb.go

```diff
@@ -1053,29 +1092,96 @@ func (bucket CouchbaseBucketGoCB) GetWithXattr(k string, xattrKey string, rv int
 
 }
 
-// Delete a document and it's associated named xattr.  Couchbase server will preserve system xattrs as part of the (CBS) tombstone when a document is deleted.
-// To remove the system xattr as well, an explicit subdoc delete operation is required.
+// Only called for Purge
 func (bucket CouchbaseBucketGoCB) DeleteWithXattr(k string, xattrKey string) error {
 
+	return bucket.deleteWithXattrInternal(k, xattrKey, nil)
+
+}
+
+// A function that will be called back after the first delete attempt but before second delete attempt
+type deleteWithXattrRaceInjection func(bucket CouchbaseBucketGoCB, k string, xattrKey string)
+
+// Delete doc + xattrs (purge).  Algorithm:
+// Try to delete body + xattr in single op
+// If fails
+// 		- Do get w/ xattr (get cas)
+// 		- Cas-safe delete just the xattr
+// If that fails with a cas error, return error (maybe race where someone resurrected doc)
+func (bucket CouchbaseBucketGoCB) deleteWithXattrInternal(k string, xattrKey string, callback deleteWithXattrRaceInjection) error {
+
 	bucket.singleOps <- struct{}{}
 	defer func() {
 		<-bucket.singleOps
 	}()
 	gocbExpvars.Add("Delete", 1)
-	removeCas, err := bucket.Bucket.Remove(k, 0)
-	if err != nil && err != gocb.ErrKeyNotFound {
+
+	LogKeys["CRUD+"] = true
+
+	LogTo("CRUD+", "DeleteWithXattr called with key: %v xattrKey: %v", k, xattrKey)
+
+	// This flag seems to work well no matter what the current document state is in
+	deleteFlags := gocb.SubdocDocFlagAccessDeleted
+
+	// Try to delete body and xattrs in single op
+
+	// If the doc exists, delete both the doc body and the xattrs in one single op
+	_, mutateErr := bucket.Bucket.MutateInEx(k, deleteFlags, gocb.Cas(0), uint32(0)).
+		RemoveEx(xattrKey, gocb.SubdocFlagXattr). // Remove the xattr
+		RemoveEx("", gocb.SubdocFlagNone).        // Delete the document body
+		Execute()
+
+	// If no error, or it was just a ErrSubDocSuccessDeleted error, we're done.
+	// ErrSubDocSuccessDeleted is a "success error" that means "operation was on a tombstoned document"
+	if mutateErr == nil || mutateErr == gocbcore.ErrSubDocSuccessDeleted {
+		LogTo("CRUD+", "No error or ErrSubDocSuccessDeleted.  We're done.")
+		return nil
+	}
+
```

Comment:
I don't think we want to continue to the fallback approach on every type of error.  What error do we get when the xattr exists but the body does not?